TS performance fixes

mvdbeek opened the pull request.

mvdbeek committed:
Avoid repeated calls to get_repo_for_repository

by creating the `hg_repo` property on the `Repository` object.  This is
only used in the toolshed, meaning we can move get_repo_for_repository
to lib/tool_shed/util/hg_util.py.

`get_repo_for_repository` takes up considerable time when hitting the
`/api/repositories/updates` route in the tool shed.  This is hit at
least whice per call
(https://gist.github.com/mvdbeek/6a58b5f4fe9489dd9a15109da74a0c86#file-gistfile1-txt-L7),
so making this a cached property should result in a nice speedup.

mvdbeek committed:
Only call repository.tip once in /api/repositories/updates

This also takes up considerable time in
https://gist.github.com/mvdbeek/6a58b5f4fe9489dd9a15109da74a0c86#file-gistfile1-txt-L6

mvdbeek committed:
Drop workaround form when mercurial wasn't py3 compatible

nsoranzo commented on lib/tool_shed/tools/tool_version_manager.py:
```diff
@@ -38,7 +38,7 @@ def get_version_lineage_for_tool(self, repository_id, repository_metadata, guid)
         """
         repository = repository_util.get_repository_by_id(self.app, repository_id)
-        repo = hg_util.get_repo_for_repository(self.app, repository=repository)
+        repo = hg_util.repository.hg_repo
```
```suggestion
        repo = repository.hg_repo
```

mvdbeek committed:
Work around sqlalchemy __init__ behavior

WEAK_HG_REPO_CACHE items will expire when Repository is garbage
collected.

nsoranzo commented on lib/tool_shed/util/hg_util.py:
```diff
@@ -121,6 +125,13 @@ def get_readable_ctx_date(ctx):
     return ctx_date
 
 
+def get_repo_for_repository(app, repository=None, repo_path=None):
```
Maybe just merge this function into `Repository.hg_repo()` ?

mvdbeek committed:
Fix hg_repo access

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on lib/tool_shed/webapp/model/__init__.py:
```diff
@@ -191,6 +194,12 @@ def __init__(self, id=None, name=None, type=None, remote_repository_url=None, ho
 
+    @property
+    def hg_repo(self):
+        if WEAK_HG_REPO_CACHE.get(self) is None:
```
```suggestion
        if not WEAK_HG_REPO_CACHE.get(self):
```

nsoranzo commented on lib/tool_shed/webapp/model/__init__.py:
```diff
@@ -191,6 +194,12 @@ def __init__(self, id=None, name=None, type=None, remote_repository_url=None, ho
+    @property
+    def hg_repo(self):
+        if WEAK_HG_REPO_CACHE.get(self) is None:
+            WEAK_HG_REPO_CACH[self] = hg_util.get_repo_for_repository(self)
```
```suggestion
            WEAK_HG_REPO_CACHE[self] = hg_util.get_repo_for_repository(self)
```

nsoranzo commented on lib/tool_shed/webapp/model/__init__.py:
```diff
@@ -191,6 +194,12 @@ def __init__(self, id=None, name=None, type=None, remote_repository_url=None, ho
+    def hg_repo(self):
+        if WEAK_HG_REPO_CACHE.get(self) is None:
+            WEAK_HG_REPO_CACH[self] = hg_util.get_repo_for_repository(self)
+        return WEAK_HG_REPO_CACH[self]
```
```suggestion
        return WEAK_HG_REPO_CACHE[self]
```

mvdbeek committed:
Use a global HgWebConfigManager instance

This avoids needing to pass app around.

mvdbeek committed:
Don't need to pass app into Repository.tip

mvdbeek committed:
Don't need to pass into Repository.is_new

mvdbeek committed:
Missed one tip() instance

mvdbeek committed:
allow_push doesn't need app

mvdbeek committed:
can_change_type doesn't need app

mvdbeek committed:
Repository.revision doesn't need app access

mvdbeek committed:
is_valid_for_type doesn't need app

mvdbeek committed:
Tweak weakreaf check

mvdbeek committed:
Lint fixes

nsoranzo commented on lib/tool_shed/webapp/model/__init__.py:
```diff
@@ -260,21 +275,21 @@ def get_tool_dependencies(self, app, changeset_revision):
-    def repo_path(self, app):
-        return app.hgweb_config_manager.get_entry(os.path.join("repos", self.user.username, self.name))
+    def repo_path(self, app=None):
+        # Keep app argument for compaibility with tool_shed_install Repository model
```
```suggestion
        # Keep app argument for compatibility with tool_shed_install Repository model
```

nsoranzo commented on lib/tool_shed/webapp/model/__init__.py:
```diff
@@ -201,14 +216,14 @@ def admin_role(self):
-    def allow_push(self, app):
-        repo = hg_util.get_repo_for_repository(app, repository=self)
-        return repo.ui.config('web', 'allow_push')
+    @property
```
I wouldn't use a property here since there is a `set_allow_push()` method which is not possible to decorate as a setter.

mvdbeek committed:
Fix cached repo usage

mvdbeek committed:
Make allow_push regular function again

mvdbeek committed:
Typo fix (thx Nicola)

mvdbeek committed:
eagerload downloadable revisions

mvdbeek committed:
eagerload downloadable_revisions where appropriate

Where appropriate means iteration over
repository.downloadable_revisions.

mvdbeek committed:
Use latest downloadable revision instead of repository.tip()

There should be no need to hit the disk in the updates route now.
We might eventually miss some local modifications that did not
create a new MetdataRevision, but I don't think managing this should
be a concern for the `updates` route.
